linkerd/service-profiles/src/http/route_request.rs

```diff
@@ -0,0 +1,130 @@
+use super::{Receiver, RequestMatch, Route};
+use crate::Profile;
+use futures::{future::ErrInto, prelude::*, ready};
+use linkerd2_error::Error;
+use linkerd2_stack::{layer, NewService, Proxy};
+use std::{
+    collections::HashMap,
+    marker::PhantomData,
+    task::{Context, Poll},
+};
+use tracing::{debug, trace};
+
+pub fn layer<M, N: Clone, R>(
+    new_route: N,
+) -> impl layer::Layer<M, Service = NewRouteRequest<M, N, R>> {
+    let default = Route::default();
+    layer::mk(move |inner| NewRouteRequest {
+        inner,
+        new_route: new_route.clone(),
+        default: default.clone(),
+        _route: PhantomData,
+    })
+}
+
+pub struct NewRouteRequest<M, N, R> {
+    inner: M,
+    new_route: N,
+    default: Route,
+    _route: PhantomData<R>,
+}
+
+pub struct RouteRequest<T, S, N, R> {
+    target: T,
+    rx: Receiver,
+    inner: S,
+    new_route: N,
+    http_routes: Vec<(RequestMatch, Route)>,
+    proxies: HashMap<Route, R>,
+    default: R,
+}
+
+impl<M: Clone, N: Clone, R> Clone for NewRouteRequest<M, N, R> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            new_route: self.new_route.clone(),
+            default: self.default.clone(),
+            _route: self._route,
+        }
+    }
+}
+
+impl<T, M, N> NewService<(Receiver, T)> for NewRouteRequest<M, N, N::Service>
+where
+    T: Clone,
+    M: NewService<(Receiver, T)>,
+    N: NewService<(Route, T)> + Clone,
+{
+    type Service = RouteRequest<T, M::Service, N, N::Service>;
+
+    fn new_service(&self, (rx, target): (Receiver, T)) -> Self::Service {
+        let inner = self.inner.new_service((rx.clone(), target.clone()));
+        let default = self
+            .new_route
+            .new_service((self.default.clone(), target.clone()));
+        RouteRequest {
+            rx,
+            target,
+            inner,
+            default,
+            new_route: self.new_route.clone(),
+            http_routes: Vec::new(),
+            proxies: HashMap::new(),
+        }
+    }
+}
+
+impl<B, T, N, S, R> tower::Service<http::Request<B>> for RouteRequest<T, S, N, R>
+where
+    B: Send + 'static,
+    T: Clone,
+    N: NewService<(Route, T), Service = R> + Clone,
+    R: Proxy<http::Request<B>, S>,
+    S: tower::Service<R::Request>,
+    S::Error: Into<Error>,
+{
+    type Response = R::Response;
+    type Error = Error;
+    type Future = ErrInto<R::Future, Error>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let mut update = None;
+        while let Poll::Ready(Some(up)) = self.rx.poll_recv_ref(cx) {
+            update = Some(up.clone());
+        }
```

Comment:
I believe the semantics of `poll_recv_ref` is that it will only return the latest update, so I'm surprised that this logic is necessary; I don't think it should ever be `Some` multiple times here.

Am I wrong about this? I'm fine with leaving this as-is regardless, but I'm curious whether this is just defensive programming or if we've actually seen issues as a result of this behavior.

Reply:
...oh, I see that this code was here before, and it just moved around in this branch. Let's not mess with it now, then.